File: python/amostragem.py
#%%
import numpy as np
import matplotlib.pyplot as plt
import sympy as sp
from numpy import pi
plt.rcParams['lines.linewidth'] = 3
plt.rcParams['font.size'] = 30
plt.rcParams['figure.figsize'] = [30,14]
#%% Function
t = sp.symbols('t')
f = sp.cos(2*np.pi*3200*t) + 0.5*sp.cos(2*np.pi*600*t) + 0.01*sp.cos(2*np.pi*300*t)
x = sp.lambdify(t,f,'numpy')
#%% Continuous
L = 10
M = 10
#Fs = int(6000/L)
Fs = int(6000*M)
Ts = 1/Fs
t_sampled = np.linspace(0,1,Fs)
x_sampled = x(t_sampled)

#%% Sampled
F = 10*6400
T = 1/F
t_continuous = np.linspace(0,1,F)
x_continuous = x(t_continuous)
# %% Plotting continuous signal
fig = plt.figure()
plt.plot(t_continuous,x_continuous,'-k')
plt.xlim([0,0.01])
plt.xticks([0, 0.002, 0.004, 0.006, 0.008, 0.01])
plt.legend(['x(t)'])
plt.xlabel('t (s)')
plt.title('Continuous time function')
plt.grid(True)
plt.show()
fig.savefig('Func_continuos.png')

#%% Plotting sampled signal
fig = plt.figure()
plt.stem(t_sampled,x_sampled,'ro')
plt.xlim([0,0.01])
plt.xticks([0, 0.002, 0.004, 0.006, 0.008, 0.01])
plt.legend(['Sampled - Fs = '+str(Fs)+' Hz'])
plt.xlabel('time (s)')
plt.title('Sampled time function')
plt.grid(True)
plt.show()
fig.savefig('Func_sampled_Fs_'+str(Fs)+'Hz.png')

#%% Plotting both signals
fig = plt.figure()
plt.plot(t_continuous,x_continuous,'-k')
plt.stem(t_sampled,x_sampled,'ro')
plt.xlim([0,0.01])
plt.xticks([0, 0.002, 0.004, 0.006, 0.008, 0.01])
plt.legend(['Continuous','Sampled - Fs = '+str(Fs)+' Hz'])
plt.xlabel('time (s)')
plt.title('Continuous and Sampled time functions')
plt.grid(True)
plt.show()
fig.savefig('Functions_Fs_'+str(Fs)+'Hz.png')

# %% FFT continuous
y_continuous = np.fft.fft(x_continuous)
y_continuous = np.abs(y_continuous)/np.max(np.abs(y_continuous))
nc = x_continuous.size
freq_continuous = np.fft.fftfreq(nc,d=T)#*T*2*pi

#%% plot FFT continuous
fig = plt.figure()
labels = ['-pi','-pi/2','0','pi/2','pi']
plt.plot(freq_continuous,y_continuous)
#plt.xticks(np.arange(-pi,pi+1,pi/2),labels)
plt.xlim([-3200-100,3200+100])
plt.xlabel('frequency (Hz)')
plt.title('Continuous Fourier transformed function')
fig.savefig('fft_continuous.png')

# %% FFT sampled
y_sampled = np.fft.fft(x_sampled)
y_sampled = np.abs(y_sampled)/np.max(np.abs(y_sampled))
n = x_sampled.size
freq_sampled = np.fft.fftfreq(n,d=Ts)#*Ts*2*pi

#%% plot FFT sampled
fig = plt.figure()
plt.stem(freq_sampled,y_sampled)
#plt.xticks(np.arange(-pi,pi+1,pi/2),labels)
#plt.xlim([-3200-100,3200+100])
plt.grid(True)
plt.xlabel('frequency (Hz)')
plt.title('Sampled Fourier transformed function')
plt.legend(['Fs = '+str(Fs)+' Hz'])
fig.savefig('fft_sampled_Fs_'+str(Fs)+'Hz.png')
#%%
fig = plt.figure()
plt.plot(freq_continuous,y_continuous,'k-')
plt.stem(freq_sampled,y_sampled,'r--')
#plt.xticks(np.arange(-pi,pi+1,pi/2),labels)
plt.xlim([-3200-100,3200+100])
plt.xlabel('frequency (Hz)')
plt.title('Continuous and Sampled Fourier transformed functions')
plt.grid(True)
plt.legend(['Continuous','Sampled - Fs = '+str(Fs)+' Hz'])
fig.savefig('fft_both_Fs_'+str(Fs)+'Hz.png')

#%% Interpolação
x_interpol = np.interp(t_sampled,t_continuous,x_continuous)
fig = plt.figure()
plt.plot(t_continuous,x_continuous,'k-')
plt.plot(t_sampled,x_sampled,'go',markersize=18)
plt.plot(t_sampled,x_interpol,'r--')
plt.xlim([0,0.01])
plt.xlabel('time (s)')
plt.title('Continuous, Sampled and Interpolated time functions')
plt.grid(True)
plt.legend(['continuous','sampled','Interpoled'])
plt.xticks([0, 0.002, 0.004, 0.006, 0.008, 0.01])
fig.savefig('func_interp_Fs_'+str(Fs)+'Hz.png')

File: matlab/amostragem.m
rng('default');
clear all;
close all;

%% Global parameters
Fs = 12000;
ns = 0:Fs;

% Continous ploting
F = 20*Fs;
n = 0:F;

% Resample
M = 2;
L = 1;
%% Sampled function  
xHandle =  @(t) cos(2*pi*3200*t) ...
            + 0.5*cos(2*pi*600*t) ...
            + 0.01*cos(2*pi*300*t);
xContinuous = xHandle(n/F);
x = xHandle(ns/Fs);

%% Plotting
figure();
fplot(xHandle,"LineWidth",1.5,'Color',[0.985 0.727 0.258])
hold on
stem(ns/Fs,x,"filled");
xlim([0 0.01]); grid on;
xticks([0 0.002 0.004 0.006 0.008 0.01])
xlabel("t [s]", "Interpreter","latex")

legend("Sinal contínuo","Amostrado","Location","northoutside",...
    "Orientation","horizontal")

%% FFT
y = fft(x);                               
nFFT = length(x);      

% double sided FFT
y0 = fftshift(y);

m = abs(y0); 
f = (-nFFT/2:nFFT/2-1)*(2*pi/nFFT); 

% FFT continuous
yC = fft(xContinuous);                               
nFFTC = length(xContinuous);      

% double sided FFT
y0C = fftshift(yC);

mC = abs(y0C); 
fC = (-nFFTC/2:nFFTC/2-1)*(2*pi*F/Fs/nFFTC); 

%
figure();
plot(fC,mC,"LineWidth",2), hold on,
plot(f,m,"LineWidth",2)
ylabel('Magnitude')
xlabel('Frequency (Hz)'),
grid;
%% Resampling
Fn = Fs/M*L;
fn = (-nFFT/2:nFFT/2-1)*(2*pi*Fs/Fn/nFFT); 
plot(fn,m,"LineWidth",2)
%xlim([-10000 10000])
legend("Sinal contínuo", "Amostrado","Reamostrado", ...
    "Location","northoutside","Orientation","horizontal")
xlim([-2*pi 2*pi])
xticks([-2*pi -1.5*pi -pi -0.5*pi 0.0 0.5*pi pi 1.5*pi 2*pi]);
xticklabels({'-2\pi','-1.5\pi','-\pi','-0.5\pi','0.0','0.5\pi','\pi','1.5\pi','2\pi'});

%% In continuous time
nn = 0:Fn;
xn = interp1(ns/Fs,x,nn/Fn);

figure();
fplot(xHandle,"LineWidth",1.5,'Color',[0.985 0.727 0.258]); hold on
stem(ns/Fs,x,"filled");
xlim([0 0.01]); grid on;
xticks([0 0.002 0.004 0.006 0.008 0.01])
xlabel("t [s]", "Interpreter","latex")

stem(nn/Fn,xn,'k');
legend("Sinal contínuo","Amostrado","Reamostrado","Location","northoutside",...
    "Orientation","horizontal")

%% Interpolation

interp = 0;
for i = 1:length(ns)
     interp = interp + x(i)*sinc((n/F-(i-1)/Fs)*Fs);
end

interpn = 0;
for i = 1:length(nn)
     interpn = interpn + xn(i)*sinc((n/F-(i-1)/Fn)*Fn);
end

figure();
fplot(xHandle,"LineWidth",3,'Color',[0.985 0.727 0.258]); hold on
plot(n/F,interp,"LineWidth",1);
xlim([0 0.01]); grid on;
xticks([0 0.002 0.004 0.006 0.008 0.01])
xlabel("t [s]", "Interpreter","latex")

plot(n/F,interpn,'k');
legend("Sinal contínuo","Amostrado","Recuperado","Location","northoutside",...
    "Orientation","horizontal")

%%
Fa = round(16e6/128/13);
% Fa = 3200*4;
na = 0:Fa;
xa = xHandle(na/Fa) + 2.5;

nBits = 10;
levels = linspace(0,5,2^nBits);
qData = discretize(xa,levels);
quant = levels(qData) - 2.5;

figure();
fplot(xHandle,"LineWidth",1.5,'Color',[0.985 0.727 0.258])
hold on
stem(na/Fa,xa-2.5,"filled");
stairs(na/Fa,quant,'k',"LineWidth",1);
xlim([0 0.006]); grid on;
xticks([0 0.002 0.004 0.006])
xlabel("t [s]", "Interpreter","latex")
ylabel("x(t)", "Interpreter","latex")
legend("Sinal contínuo","Amostrado","Quantizado","Location","northoutside",...
    "Orientation","horizontal")